File: legacy python/output/marc.py
import sys
import logging
import config
import json

from pymarc import Record, Field
import logging

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)

import harvesting

cachedPlaces = {}

def writeBatchToFile(batch, requestInfo, out):
  for place in batch:
    if("prefName" not in place):
      logger.info("Skipping place:")
      logger.info(place)
      continue;

    cachePlace(place)

    # logger.debug("Location:")
    # for(key, value in item.items()):
    #   if(key not in ["prefLocation"]):
    #     logger.debug(str(key) + ": " + str(value))

    record = Record(to_unicode=True, force_utf8=True)
    record.add_field(
      Field(
        tag = "151",
        indicators = [" ", " "],
        subfields = [
          "a", place["prefName"]["title"],
        ]),
      Field(
        tag = "40",
        indicators = [" ", " "],
        subfields = [
          "a", "iDAI.gazetteer"
        ]),
      Field(
        tag = "24",
        indicators = [" ", " "],
        subfields = [
          "a", place["gazId"],
          "2", "iDAI.gazetteer"
        ])
      )

    if "parent" in place:
      record = addParentTracing(record, place["parent"], [place["prefName"]["title"]])

    if "names" in place:
      record = addGeoTracing(record, place["names"], [place["prefName"]["title"]])

    record = customizeLeader(record)
    # logger.debug("Final marc21 record: ")
    # logger.debug(record.as_marc())

    out.write(record.as_marc())


def addParentTracing(record, parentURL, knownParents):
  parent = None

  if(parentURL in cachedPlaces):
    parent = cachedPlaces[parentURL]
    logger.debug("Using cached place: " + parent["prefName"]["title"])
  else:
    parent = harvesting.runQuery(parentURL)
    if(parent == None):
      return record

    cachePlace(parent)

  if(parent["prefName"]["title"] in knownParents):
    return record

  record.add_field(
    Field(
      tag = "551",
      indicators = [" ", " "],
      subfields = [
        "a", parent["prefName"]["title"],
        "i", "part of"
      ])
    )

  knownParents.append(parent["prefName"]["title"])

  if("parent" not in parent):
    return record # reached root https://gazetteer.dainst.org/place/2042600
  else:
    return addParentTracing(record, parent["parent"], knownParents)

def addGeoTracing(record, alternativeNames, knownNames):
  for altName in alternativeNames:
    if(altName["title"] not in knownNames):
      record.add_field(
        Field(
          tag = "451",
          indicators = [" ", " "],
          subfields = [
            "a", altName["title"]
          ])
      )
      knownNames.append(altName["title"])

  return record


# The leader encodes meta information about the marc record
# This functions sets some flags that can not be inferred automatically
# by the pymarc library:
# 'Record status', 'Type of record' and the 'Encoding level', see also
# https://www.loc.gov/marc/authority/adleader.html
def customizeLeader(record):
  record.leader = (record.leader[0:5]
    + config.formats["marc"]["options"]["record status"]
    + "z" + record.leader[7:17] + "n" + record.leader[18:])

  # logger.debug("Final leader: " + record.leader + ", length: "
  #   + str(len(record.leader)))
  return record

def cachePlace(place):
  if(place["@id"] in cachedPlaces):
    return
  elif(place["gazId"] == "2042600"):
    cachedPlaces[place["@id"]] = place
    # logger.debug("Cached place " + place["prefName"]["title"])
  elif("types" in place
    and place["types"][0] in config.formats["marc"]["options"]["cached types"]):
    cachedPlaces[place["@id"]] = place
    # logger.debug("Cached place " + place["prefName"]["title"])
  else:
    return

File: lib/gazetteer_drescher/marc.ex
defmodule GazetteerDrescher.MARC do
  @moduledoc """
  This module transforms the parsed JSON data of a place into the marc21 format.
  """
  require Logger

  alias GazetteerDrescher.MARC.Record
  alias GazetteerDrescher.MARC.Field
  alias GazetteerDrescher.Harvesting

  @doc """
  Transforms a parsed place's JSON data (as Map) into a marc21 record string.
  If the input Map is deemed invalid, an empty string is returned.

  Returns: Empty String or String with place's data in marc21.
  """
  def create_output(place) do
    if invalid?(place) do
      ""
    else
      process(place)
    end
  end

  defp process(place) do
    record = %Record{}
    record = Record.add_field(record, %Field{
      tag: 24,
      i1: "7",
      subfields: [
        a: place["gazId"],
        "2": "iDAI.gazetteer"
      ]})

    record = Record.add_field(record, %Field{
      tag: 40,
      subfields: [
        a: "iDAI.gazetteer"
      ]})

    record = Record.add_field(record, %Field{
      tag: 151,
      subfields: get_X51_subfield(place["prefName"])
    })

    record =
      record
      |> add_geo_tracing(place["names"])
      |> add_parent_tracing(place["parent"])

    Record.to_marc(record)
  end

  defp add_geo_tracing(record, nil) do
    record
  end

  defp add_geo_tracing(record, []) do
    record
  end

  defp add_geo_tracing(record, [alternative_name | tail]) do
    Record.add_field(record, %Field{
      tag: 451,
      subfields: get_X51_subfield(alternative_name)
    })
    |> add_geo_tracing(tail)
  end

  defp add_parent_tracing(record, nil) do
    record
  end

  defp add_parent_tracing(record, parent_url) do
    # cached_places = Agent.get(CachedPlaces, &(&1))
    parent =
      case :ets.lookup(:cached_places, parent_url) do
        [{^parent_url, cached_place }] ->
          # Logger.debug ~s(Using cached place: #{cached_place["prefName"]["title"]})
          cached_place
        [] ->
          response = String.replace(parent_url, "place", "doc") <> ".json"
          |> Harvesting.fetch_place

          case response do
            {:ok, data} ->
              data
            _ ->
              nil
          end
      end

    case parent do
      nil ->
        record
      _ ->
        record = record
        |> Record.add_field(%Field{
          tag: 551,
          subfields: get_X51_subfield(parent["prefName"]) ++ [ i: "part of" ]
          })
        add_parent_tracing(record, parent["parent"])
    end
  end

  defp invalid?(place) do
    cond do
      Map.has_key?(place, "prefName") == false      -> true
      place["gazId"] == nil                         -> true
      String.trim(place["gazId"]) == ""             -> true
      place["prefName"]["title"] == nil             -> true
      String.trim(place["prefName"]["title"]) == "" -> true

      true                                          -> false
    end
  end

  defp get_X51_subfield(item) do
    if item["language"] == nil or item["language"] == "" do
      [ a: item["title"] ]
    else
      [ a: item["title"], l: item["language"] ]
    end
  end
end
